File: lib/pull_request.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:nuno/pull_reqeust_commits_response.dart';
import 'package:nuno/pull_request_detail_response.dart';
import 'package:nuno/pull_request_response.dart';
import 'package:nuno/pull_request_reviews_response.dart';

part 'pull_request.freezed.dart';
part 'pull_request.g.dart';

@freezed
abstract class PullRequest with _$PullRequest {
  const factory PullRequest(
      {String title,
      String url,
      DateTime oldestCommitDate,
      DateTime createdAt,
      DateTime oldestReactionDate,
      DateTime latestApprovalDate,
      DateTime mergedAt,
      int commits,
      int additions,
      int deletions,
      int changeFiles,
      String state,
      String moduleName,
      String changeType}) = _PullRequest;

  factory PullRequest.from(PullRequestResponse pr, PullRequestDetailResponse detail,
      PullRequestCommitsResponse commits, PullRequestReviewsResponse reviews) {
    final parsedTitle = _parse_pr_title(pr.title);
    final oldestCommit = commits.oldest();
    final oldestReaction = reviews.oldestReaction();
    final latestApproval = reviews.latestApproval();

    return PullRequest(
        title: parsedTitle['title'],
        url: pr.url,
        oldestCommitDate: oldestCommit?.date,
        createdAt: pr.createdAt,
        oldestReactionDate: oldestReaction?.date,
        latestApprovalDate: latestApproval?.date,
        mergedAt: pr.mergedAt,
        commits: detail.commits,
        additions: detail.additions,
        changeFiles: detail.changeFiles,
        state: pr.state,
        moduleName: parsedTitle['module'],
        changeType: parsedTitle['type']);
  }

  factory PullRequest.fromJson(Map<String, dynamic> json) =>
      _$PullRequestFromJson(json);
}

final title_paren_pattern = RegExp(r'\(|\)');
final title_regexp = RegExp(r'(\w+)(\(\w+\))?: (.+)');

Map<String, String> _parse_pr_title(String raw_title) {
  if (!title_regexp.hasMatch(raw_title)) return {'title': raw_title};

  final result = title_regexp.firstMatch(raw_title);
  return {
    'title': result.group(3),
    'module': result.group(2)?.replaceAll(title_paren_pattern, ''),
    'type': result.group(1)
  };
}

File: src/pullRequest.ts
import type {
  PullRequestResponse,
  PullRequestDetailResponse,
  PullRequestCommitsResponse,
  PullRequestReviewsResponse,
} from './responses.js';

/**
 * 集計済みの PR 1 件。out.json に書き出され、web フロント側で読み込まれる。
 * Dart 版の freezed クラス PullRequest に対応する。
 * 日付フィールドは JSON.stringify 時に Date.toJSON()（ISO8601）でシリアライズされる。
 */
export interface PullRequest {
  title: string;
  url: string;
  oldestCommitDate: Date | null;
  createdAt: Date;
  oldestReactionDate: Date | null;
  latestApprovalDate: Date | null;
  mergedAt: Date | null;
  issueCreatedAt: Date | null;
  commits: number;
  additions: number;
  deletions: number;
  changedFiles: number;
  reviewRounds: number;
  state: string;
  moduleName: string | null;
  changeType: string | null;
}

interface ParsedTitle {
  title: string;
  module: string | null;
  type: string | null;
}

const TITLE_PAREN_PATTERN = /\(|\)/g;
const TITLE_REGEXP = /(\w+)(\(\w+\))?: (.+)/;

/** `type(module): title` 形式の PR タイトルを分解する。マッチしなければ title のみ。 */
function parsePrTitle(rawTitle: string): ParsedTitle {
  const result = TITLE_REGEXP.exec(rawTitle);
  if (result == null) {
    return { title: rawTitle, module: null, type: null };
  }
  return {
    title: result[3],
    module: result[2] != null ? result[2].replace(TITLE_PAREN_PATTERN, '') : null,
    type: result[1],
  };
}

/** 各種レスポンスを 1 件の集計済み PullRequest にまとめる。 */
export function buildPullRequest(
  pr: PullRequestResponse,
  detail: PullRequestDetailResponse,
  commits: PullRequestCommitsResponse,
  reviews: PullRequestReviewsResponse,
  issueCreatedAt: Date | null,
): PullRequest {
  const parsedTitle = parsePrTitle(pr.title);
  const oldestCommit = commits.oldest();
  const oldestReaction = reviews.oldestReaction();
  const latestApproval = reviews.latestApproval();

  return {
    title: parsedTitle.title,
    url: pr.url,
    oldestCommitDate: oldestCommit?.date ?? null,
    createdAt: pr.createdAt,
    oldestReactionDate: oldestReaction?.date ?? null,
    latestApprovalDate: latestApproval?.date ?? null,
    mergedAt: pr.mergedAt,
    issueCreatedAt,
    commits: detail.commits,
    additions: detail.additions,
    deletions: detail.deletions,
    changedFiles: detail.changedFiles,
    reviewRounds: reviews.reviewRounds(),
    state: pr.state,
    moduleName: parsedTitle.module,
    changeType: parsedTitle.type,
  };
}
